### PR TITLE
Basic feeding mechanic

### PR DESCRIPTION
**Pull request by randomPoison**
I've added the basic feeding mechanic described in #35:

- Players start with 10 balls in their food pile.
- A player's hippo eats a ball from their food pile every 0.75 seconds.
- Each ball their hippo eats gives the player 1 point.
- If the hippo tries to eat and there are no balls, the player loses and is removed from the game and the host screen.

This provides a baseline auto-removal mechanism to the game, which should fix our issues with overcrowded screens. More work needs to be done for #35, but this is working well enough to deploy it while we add more functionality.

### CODE REVIEW

File: src/api.rs [hunk 1]
```diff
@@ -5,54 +5,58 @@ use rocket::http::Status;
...
+    let id = player_id_generator.next_id();
     let username = game::generate_username();
+    let balls = 10;
 
-    // Add the username to the Usernames map
-    {
-        let mut usernames = usernames.lock().expect("Usernames mutex was poisoned");
-        let old = usernames.insert(player_id, username.clone());
-        assert_eq!(None, old, "Player ID was registered twice");
+    let player = Player {
```
Comment (wufire) [inline src/api.rs:32]: Hello player object. I did not expect to so you so soon

Comment (randomPoison) [reply src/api.rs:32]: Oh yeah, I forgot to mention that I removed all the separate maps for scoreboard and usernames and such. They were making the code harder to work with and were a speculative optimization that we probably don't need yet in practice. So I've unified everything into `Player` for now, and we can split things back out if accessing the player map becomes a bottleneck.

File: src/api.rs [hunk 2]
```diff
@@ -78,36 +82,29 @@ pub struct FeedPlayerResponse {
...
 {
     let payload = payload.into_inner();
-    let player_id = payload.player;
+    let id = payload.player;
 
     // Add 1 to the player's score, returning the new score.
-    let score = {
-        let mut scoreboard = scoreboard.lock().expect("Scoreboard mutex was poisoned");
+    let balls = {
+        let mut players = players.write().expect("Players were poisoned");
```
Comment (wufire) [inline src/api.rs:94]: The players were poisoned with cyanide.

Comment (wufire) [reply src/api.rs:94]: Haha, so a review is much more serious than a `comment`. Neat!

File: src/api.rs [hunk 3]
```diff
@@ -78,36 +82,29 @@ pub struct FeedPlayerResponse {
...
 
         // Get the player's current score, or return an `InvalidPlayer` error if it's not in
         // the scoreboard.
-        let score = scoreboard
-            .get_mut(&player_id)
-            .ok_or(Error::InvalidPlayer(player_id))?;
-        *score += 1;
-        *score
```
Comment (wufire) [inline src/api.rs:98]: In the old version here, what did the `*` do?

Comment (wufire) [reply src/api.rs:98]: I know I could just look that up, but now it's part of a code review.

File: src/api.rs [hunk 4]
```diff
@@ -119,24 +116,40 @@ pub struct PlayersResponse {
...
-    // mutex for too long.
-    let scoreboard = scoreboard.lock().expect("Scoreboard mutex was poisoned").clone();
-    let mut usernames = usernames.lock().expect("Usernames mutex was poisoned").clone();
-
-    let players = usernames.drain()
-        .map(|(id, username)| {
-            let score = *scoreboard.get(&id).expect("Score for player was not in scoreboard");
-            PlayerData { id, username, score }
+pub fn get_players(players: State<PlayerMap>) -> JSON<PlayersResponse> {
+    let players = players.read().expect("Player map was poisoned!");
```
Comment (wufire) [inline src/api.rs:144]: Why were we `lock`ing before and now we are `read`ing?

File: src/game.rs [hunk 5]
```diff
@@ -145,4 +146,58 @@ pub struct PlayerData {
...
+            let now = Instant::now();
+            {
+                let mut players = players.write().expect("Hippo map was poisoned!");
+                players.retain(|&id, player| {
+                    // Ignore hippos that are not ready to eat.
+                    if now < player.next_eat_time { return true; }
+
+
+                    // Try to eat a ball. If there's one for the hippo to eat, we get a point.
+                    // Otherwise, the hippo is le dead.
```
Comment (wufire) [inline src/game.rs:171]: ☠️

File: www/host.js [hunk 6]
```diff
@@ -56,22 +57,31 @@ socket.onmessage = (event) => {
...
 
         // Updated the local score for the player.
         hippo.player.score = info.score;
+        hippo.player.balls = info.balls;
 
         // Animate the hippo head to match the score increase. The direction of the chomp animation
         // depends on the side of the screen that the hippo is on.
         let element = document.getElementById(hippo.player.id);
-        switch (hippo.side) {
+        switch (hippo.side.side) {
```
Comment (wufire) [inline www/host.js:84]: `side.side`?

**Review by wufire — CHANGES_REQUESTED** [review]

File: src/api.rs (hunk 1 above)
Comment (wufire) [reply src/api.rs:32]: I think this is the correct path for now. Easier to think of conceptually, too.

**Review by wufire — APPROVED** [review]

File: src/api.rs (hunk 3 above)
Comment (randomPoison) [reply src/api.rs:98]: So `score` as a `&usize`, meaning a pointer to a `usize`. To change the `usize` that it points to, you have to dereference the pointer, which is what `*` does. If you didn't have the `*`, then Rust would think you're trying to change the pointer itself.

File: src/api.rs (hunk 4 above)
Comment (randomPoison) [reply src/api.rs:144]: Before we were using a `Mutex`, now we're using `RwLock`. Mutex ensures that only one thread can access the data, and that thread always has mutable access once it has acquired the lock. `RwLock` has both `read()` method, which allows multiple threads to read the data at the same time, and a `write()` method, which behaves the same as `lock()` did.

Switching to `RwLock` give us a little bit of leeway to have multiple threads access the data in parallel if they don't need to mutate it.

File: www/host.js (hunk 6 above)
Comment (randomPoison) [reply www/host.js:84]: `hippo.side` is an object: `{ side: number, array: Hippo[] }`. That could be renamed to `hippo.side.sideId` or something to make it clearer.

File: src/api.rs (hunk 3 above)
Comment (wufire) [reply src/api.rs:98]: Ooo, does rust allow you to manipulate the pointers directly?

File: src/api.rs (hunk 4 above)
Comment (wufire) [reply src/api.rs:144]: Ah, yeah. I remember talking about this, now.

File: src/api.rs (hunk 3 above)
Comment (randomPoison) [reply src/api.rs:98]: Yuppers. Rust is a lot like C/C++ in that most of the fundamental parts of the language are very thin abstractions over how the hardware works. Within an `unsafe` block, you can manipulate raw memory in the same way you would C/C++. Normally, though, Rust restricts you to only performing safe operations, even though you're always working with pointers when you use reference/borrows.

In Rust terms, though, `&usize` is a "reference to a `usize`" or "a `usize` borrow". Under the hood it's still functionally a pointer, hence why you need to manually dereference it, but Rust also "raw pointers", which allow you to do all the unsafe pointer hackery you can do in C/C++.

Comment (wufire) [reply src/api.rs:98]: Ooo, a `safe`/`unsafe` mode is a good idea. Allows both the dangerous superpowers and the reasonable limitations.

**Comment by randomPoison** [comment]
Fun fact: C# also has an `unsafe` block that let's you handle raw pointers.

**Pull request merged**